File: typescript/src/crypto.ts
/**
 * Cryptographic operations for ONP: X25519 ECDH, HKDF-SHA256, and ChaCha20-Poly1305 AEAD.
 */

import * as crypto from 'crypto';
import { HANDSHAKE_NONCE_SIZE, NONCE_SIZE, PUBLIC_KEY_SIZE } from './constants';

const X25519_SPKI_PREFIX = Buffer.from('302a300506032b656e032100', 'hex');

export interface SessionKeys {
  clientWriteKey: Buffer;
  serverWriteKey: Buffer;
  opcodeSeed: Buffer;
}

export class KeyExchange {
  private privateKey: crypto.KeyObject;
  public publicKey: Buffer;

  constructor() {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('x25519');
    this.privateKey = privateKey;
    const spkiDer = publicKey.export({ type: 'spki', format: 'der' });
    this.publicKey = spkiDer.subarray(spkiDer.length - PUBLIC_KEY_SIZE);
  }

  /**
   * Computes the Diffie-Hellman shared secret with the peer's raw 32-byte public key.
   */
  public computeSharedSecret(peerRawPublic: Uint8Array): Buffer {
    const peerDer = Buffer.concat([X25519_SPKI_PREFIX, Buffer.from(peerRawPublic)]);
    const peerPubKeyObj = crypto.createPublicKey({
      key: peerDer,
      format: 'der',
      type: 'spki',
    });

    return crypto.diffieHellman({
      privateKey: this.privateKey,
      publicKey: peerPubKeyObj,
    });
  }
}

/**
 * Derives symmetric session keys and polymorphic opcode seed from ECDH shared secret.
 */
export function deriveSessionKeys(
  sharedSecret: Uint8Array,
  clientNonce: Uint8Array,
  serverNonce: Uint8Array
): SessionKeys {
  const salt = Buffer.concat([Buffer.from(clientNonce), Buffer.from(serverNonce)]);
  const okm = crypto.hkdfSync(
    'sha256',
    Buffer.from(sharedSecret),
    salt,
    Buffer.from('ONP-v1-SESSION-KEYS'),
    96
  );

  const okmBuf = Buffer.from(okm);
  return {
    clientWriteKey: okmBuf.subarray(0, 32),
    serverWriteKey: okmBuf.subarray(32, 64),
    opcodeSeed: okmBuf.subarray(64, 96),
  };
}

/**
 * Encrypts plaintext using ChaCha20-Poly1305 with Associated Data (AAD).
 * Returns Buffer containing [Ciphertext || 16-byte MAC Tag].
 */
export function encryptAead(
  key: Uint8Array,
  nonce: Uint8Array,
  aad: Uint8Array,
  plaintext: Uint8Array
): Buffer {
  const cipher = crypto.createCipheriv(
    'chacha20-poly1305',
    Buffer.from(key),
    Buffer.from(nonce),
    { authTagLength: 16 }
  );

  cipher.setAAD(Buffer.from(aad), { plaintextLength: plaintext.length });
  const ciphertext = cipher.update(Buffer.from(plaintext));
  const final = cipher.final();
  const tag = cipher.getAuthTag();

  return Buffer.concat([ciphertext, final, tag]);
}

/**
 * Decrypts ciphertext and verifies 16-byte Poly1305 MAC tag with Associated Data (AAD).
 */
export function decryptAead(
  key: Uint8Array,
  nonce: Uint8Array,
  aad: Uint8Array,
  ciphertextAndTag: Uint8Array
): Buffer {
  if (ciphertextAndTag.length < 16) {
    throw new Error('Ciphertext too short: missing 16-byte Poly1305 MAC tag');
  }

  const tag = ciphertextAndTag.subarray(ciphertextAndTag.length - 16);
  const ciphertext = ciphertextAndTag.subarray(0, ciphertextAndTag.length - 16);

  const decipher = crypto.createDecipheriv(
    'chacha20-poly1305',
    Buffer.from(key),
    Buffer.from(nonce),
    { authTagLength: 16 }
  );

  decipher.setAAD(Buffer.from(aad), { plaintextLength: ciphertext.length });
  decipher.setAuthTag(Buffer.from(tag));

  const plaintext = decipher.update(Buffer.from(ciphertext));
  const final = decipher.final();

  return Buffer.concat([plaintext, final]);
}

File: rust/onp-core/src/crypto.rs
//! Authenticated Encryption (ChaCha20-Poly1305) and Ephemeral Key Exchange (X25519 + HKDF-SHA256).

use chacha20poly1305::{
    aead::{Aead, KeyInit, Payload},
    ChaCha20Poly1305, Key, Nonce,
};
use hkdf::Hkdf;
use rand::rngs::OsRng;
use sha2::Sha256;
use x25519_dalek::{EphemeralSecret, PublicKey};

use crate::constants::{HANDSHAKE_NONCE_SIZE, NONCE_SIZE, PUBLIC_KEY_SIZE};
use crate::errors::{OnpError, OnpResult};

/// Ephemeral key exchange bundle holding the local secret and public key.
pub struct KeyExchange {
    secret: Option<EphemeralSecret>,
    pub public_key: [u8; PUBLIC_KEY_SIZE],
}

impl KeyExchange {
    /// Generates a fresh ephemeral Curve25519 keypair from system CSPRNG.
    pub fn new() -> Self {
        let secret = EphemeralSecret::random_from_rng(OsRng);
        let public_key = PublicKey::from(&secret).to_bytes();
        Self {
            secret: Some(secret),
            public_key,
        }
    }

    /// Performs Diffie-Hellman key exchange with peer's public key.
    /// Consumes the ephemeral secret so it can never be reused.
    pub fn complete(mut self, peer_public_bytes: &[u8; PUBLIC_KEY_SIZE]) -> OnpResult<[u8; 32]> {
        let secret = self.secret.take().ok_or_else(|| {
            OnpError::KeyExchangeFailed("Ephemeral secret already consumed".into())
        })?;
        let peer_public = PublicKey::from(*peer_public_bytes);
        let shared = secret.diffie_hellman(&peer_public);
        Ok(*shared.as_bytes())
    }
}

/// Derived 32-byte cryptographic keys and opcode permutation seed for an ONP session.
#[derive(Clone)]
pub struct SessionKeys {
    pub client_write_key: [u8; 32],
    pub server_write_key: [u8; 32],
    pub opcode_seed: [u8; 32],
}

/// Derives symmetric session keys and polymorphic opcode seed from ECDH shared secret.
pub fn derive_session_keys(
    shared_secret: &[u8; 32],
    client_nonce: &[u8; HANDSHAKE_NONCE_SIZE],
    server_nonce: &[u8; HANDSHAKE_NONCE_SIZE],
) -> OnpResult<SessionKeys> {
    let mut salt = [0u8; HANDSHAKE_NONCE_SIZE * 2];
    salt[..HANDSHAKE_NONCE_SIZE].copy_from_slice(client_nonce);
    salt[HANDSHAKE_NONCE_SIZE..].copy_from_slice(server_nonce);

    let hk = Hkdf::<Sha256>::new(Some(&salt), shared_secret);
    let mut okm = [0u8; 96];
    hk.expand(b"ONP-v1-SESSION-KEYS", &mut okm)
        .map_err(|e| OnpError::KeyExchangeFailed(format!("HKDF expand error: {e}")))?;

    let mut client_write_key = [0u8; 32];
    let mut server_write_key = [0u8; 32];
    let mut opcode_seed = [0u8; 32];

    client_write_key.copy_from_slice(&okm[0..32]);
    server_write_key.copy_from_slice(&okm[32..64]);
    opcode_seed.copy_from_slice(&okm[64..96]);

    Ok(SessionKeys {
        client_write_key,
        server_write_key,
        opcode_seed,
    })
}

/// Encrypts plaintext using ChaCha20-Poly1305 with Associated Data (AAD).
pub fn encrypt_aead(
    key: &[u8; 32],
    nonce: &[u8; NONCE_SIZE],
    aad: &[u8],
    plaintext: &[u8],
) -> OnpResult<Vec<u8>> {
    let cipher = ChaCha20Poly1305::new(Key::from_slice(key));
    let nonce_obj = Nonce::from_slice(nonce);

    let payload = Payload {
        msg: plaintext,
        aad,
    };

    cipher
        .encrypt(nonce_obj, payload)
        .map_err(|_| OnpError::AuthenticationFailed)
}

/// Decrypts ciphertext and verifies Poly1305 MAC tag with Associated Data (AAD).
pub fn decrypt_aead(
    key: &[u8; 32],
    nonce: &[u8; NONCE_SIZE],
    aad: &[u8],
    ciphertext_and_tag: &[u8],
) -> OnpResult<Vec<u8>> {
    let cipher = ChaCha20Poly1305::new(Key::from_slice(key));
    let nonce_obj = Nonce::from_slice(nonce);

    let payload = Payload {
        msg: ciphertext_and_tag,
        aad,
    };

    cipher
        .decrypt(nonce_obj, payload)
        .map_err(|_| OnpError::AuthenticationFailed)
}
